presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestGroupInnerJoinsByConnectorRuleSet.java

```diff
@@ -0,0 +1,490 @@
+                newBigintVariable("b1"), new TestingColumnHandle("b1"),
+                newBigintVariable("b2"), new TestingColumnHandle("b2"));
+        List<VariableReferenceExpression> tableOutputVars2 = ImmutableList.of(newBigintVariable("b1"), newBigintVariable("b2"));
+
+        JoinTableInfo tableInfo1 = new JoinTableInfo(new TestingMetadata.TestingTableHandle(new SchemaTableName(TEST_SCHEMA, TEST_TABLE)),
+                tableAssignments1, tableOutputVars1);
+        JoinTableInfo tableInfo2 = new JoinTableInfo(new TestingMetadata.TestingTableHandle(new SchemaTableName(TEST_SCHEMA, TEST_TABLE)),
+                tableAssignments2, tableOutputVars2);
+
+        joinTableSet1.add(tableInfo1);
+        joinTableSet1.add(tableInfo2);
+        JoinTableSet tableHandleSet1 = new JoinTableSet(joinTableSet1);
+        TableHandle tableHandle1 = new TableHandle(
+                new ConnectorId(CATALOG_SUPPORTING_JOIN_PUSHDOWN),
+                tableHandleSet1,
+                TestingTransactionHandle.create(),
+                Optional.empty());
+
+        Set<JoinTableInfo> joinTableSet2 = new HashSet<>();
+        Map<VariableReferenceExpression, ColumnHandle> tableAssignments3 = ImmutableMap.of(
+                newBigintVariable("c1"), new TestingColumnHandle("c1"),
+                newBigintVariable("c2"), new TestingColumnHandle("c2"));
+        List<VariableReferenceExpression> tableOutputVars3 = ImmutableList.of(newBigintVariable("c1"), newBigintVariable("c2"));
+
+        Map<VariableReferenceExpression, ColumnHandle> tableAssignments4 = ImmutableMap.of(
+                newBigintVariable("d1"), new TestingColumnHandle("d1"),
+                newBigintVariable("d2"), new TestingColumnHandle("d2"));
+        List<VariableReferenceExpression> tableOutputVars4 = ImmutableList.of(newBigintVariable("d1"), newBigintVariable("d2"));
+
+        JoinTableInfo table2Info1 = new JoinTableInfo(new TestingMetadata.TestingTableHandle(new SchemaTableName(TEST_SCHEMA, TEST_TABLE)),
+                tableAssignments3, tableOutputVars3);
+        JoinTableInfo table2Info2 = new JoinTableInfo(new TestingMetadata.TestingTableHandle(new SchemaTableName(TEST_SCHEMA, TEST_TABLE)),
+                tableAssignments4, tableOutputVars4);
+
+        joinTableSet2.add(table2Info1);
+        joinTableSet2.add(table2Info2);
+        JoinTableSet tableHandleSet2 = new JoinTableSet(joinTableSet2);
+        TableHandle tableHandle2 = new TableHandle(
+                new ConnectorId(OTHER_CATALOG_SUPPORTING_JOIN_PUSHDOWN),
+                tableHandleSet2,
+                TestingTransactionHandle.create(),
+                Optional.empty());
+
+        assertGroupInnerJoinsByConnectorRuleSet()
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.join(
+                                        INNER,
```

Comment:
Can we mix this up - instead of having a first level joins between the same catalog table, create a first level join between different catalogs.

Use a join clause that allows for a JOIN between all tables like `table1.col1 = table2.col1 AND table2.col1 = table3.col1 and table3.col1 = table4.col1 and table4.col1 = table1.col1`